File: src/main/kotlin/fr/jhelp/kotlinLigthTools/atomic/AtomicBool.kt
package fr.jhelp.kotlinLigthTools.atomic

import fr.jhelp.kotlinLight.Mutex

class AtomicBool
{
    private val mutex: Mutex
    private var value: Boolean

    constructor(value: Boolean)
    {
        this.mutex = Mutex()
        this.value = value
    }

    fun get(): Boolean
    {
        var value = false
        this.mutex.safeExecute { value = this.value }
        return value
    }

    fun set(value: Boolean)
    {
        this.mutex.safeExecute { this.value = value }
    }

    fun getAndSet(newValue: Boolean): Boolean
    {
        var value = false

        this.mutex.safeExecute {
            value = this.value
            this.value = newValue
        }

        return value
    }

    fun compareAndSet(excepted: Boolean, newValue: Boolean): Boolean
    {
        var done = false

        this.mutex.safeExecute {
            if (this.value == excepted)
            {
                this.value = newValue
                done = true
            }
        }

        return done
    }
}

File: src/main/swift/Sources/swift/fr/jhelp/kotlinLigthTools/atomic/AtomicBool.swift


public class AtomicBool {

    private let mutex: Mutex
    private var value: Bool

    public init(_ value: Bool){
        self.mutex = Mutex()
        self.value = value
    }

    public func get() -> Bool
    {
        var value = false
        self.mutex.safeExecute { value = self.value }
        return value
    }

    public func set(_ value: Bool)
    {
        self.mutex.safeExecute { self.value = value }
    }

    public func getAndSet(_ newValue: Bool) -> Bool
    {
        var value = false

        self.mutex.safeExecute {
            value = self.value
            self.value = newValue
        }

        return value
    }

    public func compareAndSet(_ excepted: Bool, _ newValue: Bool) -> Bool
    {
        var done = false

        self.mutex.safeExecute {
            if self.value == excepted
            {
                self.value = newValue
                done = true
            }
        }

        return done
    }
}
